### PR TITLE
Test feature macro OpenCL 3.0

### LINKED ISSUE #755: OpenCL 3.0 feature macros CTS test
I am going to prepare CTS test  to validate feature macros in OpenCL 3.0
**Where:** Inside of test_compiler suite
**How:** Similar to test_image_macro.cpp . The test will verify if particular param_names values returned by clDeviceInfo are equal to values that corresponds to turn on/off feature. Then appropriate kernel will be executed to validate feature turn on/off
**Test scope:**
- [x] __opencl_c_3d_image_writes
- [x] __opencl_c_atomic_order_acq_rel
- [x] __opencl_c_atomic_order_seq_cst
- [x] __opencl_c_atomic_scope_device
- [x] __opencl_c_atomic_scope_all_devices
- [x] __opencl_c_device_enqueue
- [x] __opencl_c_generic_address_space
- [x] __opencl_c_pipes
- [x] __opencl_c_program_scope_global_variables
- [x] __opencl_c_read_write_images
- [x] __opencl_c_subgroups
- [x] __opencl_c_work_group_collective_functions

### PR DESCRIPTION
**Pull request by gwawiork**
Fixes #755

### CODE REVIEW

File: test_conformance/compiler/main.cpp
Comment (kpet) [inline test_conformance/compiler/main.cpp:90]: This looks incorrect to me. Isn't the min required version 2.0?

**Comment by gwawiork** [comment]
> I suggest using tables and a loop to reduce code duplication and improve maintainability.

Do you suggest have one command line with multiple subcases ? I would stay with separate list of subcases

**Comment by kpet** [comment]
I think the test should loop over {feature macro, device info checks} pairs instead of having the code duplicated many times over. This could be a single test from the framework's PoV. That being said, I agree it would be nice to report individual results but the harness doesn't give us that ability at the moment. If we have to choose, my vote goes to making the code shorter and more generic.

**Comment by gwawiork** [comment]
Refactored - use template function. Test cases are separated.

**Comment by gwawiork** [comment]
I will refactor test to have one test case. 
The step with comparing content of CL_DEVICE_OPENCL_C_FEATURES will be added

**Comment by kpet** [comment]
@gwawiork This now needs rebasing.

**Comment by gwawiork** [comment]
Rebased,

File: test_conformance/compiler/main.cpp
Comment (kpet) [inline test_conformance/compiler/main.cpp:100]: Please add these two tests back :).

File: test_conformance/compiler/test_feature_macro.cpp
Comment (kpet) [inline test_conformance/compiler/test_feature_macro.cpp:20]: Suggest using C++11 raw literals to avoid having to explicitly manage line endings.

File: test_conformance/compiler/test_feature_macro.cpp
Comment (kpet) [inline test_conformance/compiler/test_feature_macro.cpp:41]: This should ideally be split into separate functions to deal with the different kinds of queries. You could have several specialised `test_feature_macro` functions (overloads or different names) that share the compiler and comparison part. This is not strictly required but would make the code a bit easier to read IMHO.

File: test_conformance/compiler/test_feature_macro.cpp
Comment (kpet) [inline test_conformance/compiler/test_feature_macro.cpp:128]: Maybe add an `else` here for readability.

File: test_conformance/compiler/test_feature_macro.cpp
Comment (kpet) [inline test_conformance/compiler/test_feature_macro.cpp:393]: Need to add the following features:
- `__opencl_c_images`
- `__opencl_c_fp64`
- `__opencl_c_int64`

**Review by kpet — CHANGES_REQUESTED** [review]
Looks generally good. The code could be made a bit easier to read and few macros added after the initial spec need to be covered as well.

**Comment by gwawiork** [comment]
@kpet I think all you requests have been addressed in last commits

File: test_conformance/compiler/test_feature_macro.cpp
Comment (gwawiork) [reply test_conformance/compiler/test_feature_macro.cpp:332]: ok thanks, fixed.

**Comment by bashbaug** [comment]
Waiting for one final review but aiming to merge soon.  Ping @kpet

**Review by kpet — DISMISSED** [review]
Not a thorough review but looks good. As discussed on the 2020/09/22 teleconference, it's now more valuable to merge this as a base change and iterate on top if need be.

**Comment by kpet** [comment]
@lakshmih @bcalidas Any objections to this getting merged? AFAICT your comments have been addressed.

File: test_conformance/compiler/main.cpp
Comment (alycm) [inline test_conformance/compiler/main.cpp:92]: This should be 3.0, right?

```suggestion
    ADD_TEST_VERSION(features_macro, Version(3, 0)),
```

File: test_conformance/compiler/test_feature_macro.cpp
Comment (alycm) [inline test_conformance/compiler/test_feature_macro.cpp:25]: The point of `ERROR;` is to cause a compiler failure?  Using an error directive seems more explicit.

```suggestion
            #error Expected feature macro was not defined
```

File: test_conformance/compiler/test_feature_macro.cpp
Comment (alycm) [inline test_conformance/compiler/test_feature_macro.cpp:34]:
```suggestion
            #error Feature macro was defined when it was not expected
```

File: test_conformance/compiler/test_feature_macro.cpp
Comment (alycm) [inline test_conformance/compiler/test_feature_macro.cpp:691]: Because this checks the entire range of `vec_device_feature_names` and compares each element of `vec_to_cmp` against it, if `vec_to_cmp` is the longer list then then will pass even though the lists clearly aren't equal.

Swapping the names doesn't fix that because what if `vec_device_feature_names` is longer.

An example of genuine output that clearly contains a problem, but still PASSED:

```
Comparison list of features: CL_DEVICE_OPENCL_C_FEATURES vs API/compiler queries.
Empty list of CL_DEVICE_OPENCL_C_FEATURES returned by clGetDeviceInfo on this device.
Comparison list of features - passed
Supported features based on CL_DEVICE_OPENCL_C_FEATURES API query:

Supported features based on queries to API/compiler :
__opencl_c_fp64
__opencl_c_int64
features_macro passed
PASSED sub-test.
PASSED test.
```

If I switch this to

```suggestion
    cl_bool result = vec_device_feature_names == vec_to_cmp;
```

the test then fails as you would expect.

**Review by alycm — COMMENTED** [review]
LGTM, I think the 2.0 vs 3.0 point is the only point that I think is wrong, everything else is a quality point that could be addressed in a new PR.

**Comment by gwawiork** [comment]
@alycm Thanks . Version fixed other issues I will open new github Issue

**Review by alycm — APPROVED** [review]
Thanks, LGTM, remaining comments moved to https://github.com/KhronosGroup/OpenCL-CTS/issues/979.

**Comment by alycm** [comment]
I'm going to merge this, multiple people have looked at this and I believe all issues have been addressed.  If I'm wrong please file issues for any outstanding point.

**Pull request merged**
